### DISCUSSION TITLE
Bring in Kasa Devices

### DISCUSSION THREAD

**Discussion by jamesonuk** [petretiandrea/home-assistant-tapo-p100#562] [Ideas]
Since the Tapo app updated the other day it has brought across a few Kasa devices I had setup.

Most deice are supported by the inbuilt HA TP Link Kasa integration but the one particular thing of interest is the KH100 Hub (oddly it looks identical to the H100 Tapo hub 😜 but whilst other devices are similar this even shares exactly the same firmware version so I will assume they are in fact the same devices running the same software).   If you try and pair devices to this KH100 hub then the Tapo app gives you the same options as the Tapo H100 hub and the Kasa app gives you the same list but also adds in the KE100 Radiator Value

I tried to add the KH100 hub via this integration but I got device not supported error.

The main thing of interest here is the Kasa KE100 radiator valves.  These are not supported by the inbuilt Kasa integration (which looks primarily to be because they need go pair via the hub which is almost certainly using the Tapo protocols) but can't be added added via the Tapo app (and although the hub appears in the Tapo app, the valve doesn't).

I am really hoping that it might be possible to integrate these valves into HA and this seems to be a possible route...

**Comment by jamesonuk** [comment]
In fact just changing SUPPORTED_HUB_DEVICE_MODEL in const.pyt to be a list containing KH100 and changing the == in __inst__.py to be an in the integration picked up the hub and also the valve....
![image](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/1040621/5d24b501-e3cd-4019-8867-2e890f6e03f5)

**Reply by Seliaks65** [reply]
Hi @jamesonuk @delatt @els8791 , I have a similar setup with Kasa app with kh100 hub and 3 ke100 trvs, and tapo app with 2 smart plugs, plus two temp sensors which show up on both apps. I have edited const.py as suggested but am unable to locate the inst.py file - and my kh100 hub wont set up in the integration, so the trvs and the temp sensors have not appeared in the integration yet, but I added the integration a couple of months ago and the two smart plugs appeared ok, so I got it to work at least for those. Any ideas what I'm doing wrong? Thanks, Tom

**Reply by jamesonuk** [reply]
@Seliaks65 the integration has been updated so that the hub works out the box.   In fact even better than that, the TRVs have been added and seem to work pretty well.

![image](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/1040621/76e3d72a-efac-4b07-bac9-4f078c2b0c0f)

so you should just need to setup the hub in Tapo app then setup integration in HA and add devices.
If it isn't working that you probably need raise an issue here but it is working for me

**Reply by Seliaks65** [reply]
@jamesonuk thanks - I was on an old version of the integration, I had some trouble getting it updated, but it's now brought all the Tapo and Kasa devices into HA - thanks for your very prompt response and assistance.

**Comment by delatt** [comment]
Any luck with this? Interested in the Kasa TRVs but unsure how well they'll work with HA. Any info would be appreciated :)

**Comment by els8791** [comment]
Many thanks @jamesonuk
I have these exact components and managed to add the KH100 Hub to HA.
The KE100 are recognized 7 out of 9 but I have no control, scene or entity.
I really hope that support will be possible :)

![Bildschirmfoto vom 2023-10-31 23-10-44](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/104236321/a0988d62-415e-4614-b203-7d4b88868b37)

![Bildschirmfoto vom 2023-10-31 23-16-50](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/104236321/41d08ab8-6663-4c00-a590-c8c2851ff863)

**Comment by ghervasze** [comment]
@jamesonuk please, please help me add the KH100 devices also. I am at wit's end - been trying to do it for the last 3 days to no avail. I am actually trying to add some temp sensors, that is T310 and T315. Managed to add the hub via the custom HACS integration but no devices (temp and humidity sensors) to it. They should work as they are supported per the integration documentation. I appreciate any help or at least where to look. Did I mention I am new to all this ..? 

![integration readme](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/167068000/a7937677-c6a1-4a7e-9157-de414ef7d29e)


![hub config](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/167068000/38a0099d-3c18-4e49-8471-9529ab7af6fa)

![devices](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/167068000/04152a87-57c3-4375-a7ef-c13b40d980e6)

**Reply by jamesonuk** [reply]
Unfortunately I gave up on the TP Link TRVs so no longer have the hub.
IIRC once the hub appeared in the Tapo app then everything worked OK and it brought in the TRV and temperature sensors I had linked to it but I can no longer check

**Reply by ghervasze** [reply]
Appreciate the very quick response. Any pointers as to where to ask for further assistance?

**Reply by jamesonuk** [reply]
Probably raise a new issue here.   You have obviously managed to get the hub in but you are missing the things paired to it?    Just check you can actually see them in the Tapo app and is so raise an issue

**Reply by ghervasze** [reply]
Thanks. I will do that. Have a good one.

**Reply by Seliaks65** [reply]
@ghervasze - I found that the recent updates to the HACSTapo integration killed my T310 temp sensors off and TRVs and made them unavailable. In the end, the only way I found to fix it was to redownload version 3.0.0 using HACs and restart Home Assistant and the sensors and TRVs came back. Not sure what's going on there with the 3.1.x versions. Info on how to redownload older packages is here [https://community.home-assistant.io/t/how-do-i-down-grade-a-hacs-loaded-custom-integration/683859](url)

**Reply by ghervasze** [reply]
@Seliaks65 thank you so much for the suggestion. I will try to do that and hope for the best.

**Reply by metinanac** [reply]
Hi, @Seliaks65 and @ghervasze

Would you be so kind to tell me how to download the version 3.0.0? Do you have the link for the repository please?

**Reply by ghervasze** [reply]
@metinanac 
Hello. You don't have to add the repository manually. You simply go to HACS, search for tapo, click on the 3 dots at the right of the integration name, redownload, select version from the dropdown, wait for the download to complete, restart and that's it. After restart your displayed version for the tapo integration  on the integrations page should be 3.0.0 if it all went okay.


![Screenshot 2024-05-07 173644](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/167068000/77bf2b7d-eb9e-4655-86e4-372a930e4676)

![Screenshot 2024-05-07 173750](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/167068000/91600b74-7ab1-465f-9eca-15508523fa7f)

![Screenshot 2024-05-07 173454](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/167068000/00a002b5-10f6-470e-928d-3548e9251554)


![Screenshot 2024-05-07 174029](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/167068000/4468b309-023e-436f-a695-6b3ecc5fc56b)

**Comment by rorygallagher2024** [comment]
Just adding: I was running an older version of this integration and had recently lost my KE100 devices and tp-link motion sensor from the Tapo Hub which just show as 'unavailable' within home assistant. I upgraded to the latest version (3.1.2) but the problem persisted. I have now downgraded to 3.0 using the instructions above and the thermostats and motion sensor are showing again.

Two additional things I observed: 
- When I tried to downgrade using HACS, I would select version 3.0 from the dropdown but it kept reverting to 3.1.2 in the dropdown label. This is just a little bug as when I clicked 'install' it did actually install 3.0 as requested.
- The latest version of my Tapo android app has also recently lost my thermostats which is really strange, and I'm not the only person to have noticed this as I have found two other mentions in the Tapo/TP-link forums. The solution to that was to downgrade the app too.

Thanks all for the info!

**Reply by ghervasze** [reply]
Yep, the first bug with the version number remaining the same I saw too. However it's more of a cosmetic issue rather than functional.
I have not had the android problem you described. It all still works, regardless of the version. I am not sure HA can affect the app on your android. These function separately, you can have automations in your tapo app that work in parallel with the HA automations.

**Comment by metinanac** [comment]
Many thanks to you both. I was experiencing the same problem with KE100 devices but now got them all back by donwgrading the version to 3.0.0 by following your instructions.
Much appreciated!

**Reply by ghervasze** [reply]
No problem, glad it worked out. Hopefully the next update will fix the issue.
